Add tests for snapshotting and compaction

Zelldon opened the pull request:
## Description

Improved further the `RaftRule` such that we are able to trigger snapshotting, which will create snapshots on all nodes. After the snapshot is taken the snapshot listeners are called, which trigger compaction. This setup reflects now how our system works, but it is much easier to test I think.

Normally we do a snapshot on the Leader and replicate that. The snapshot replication call at the end also `SnapshotStore#newSnapshot`. With the current test approach we were able to simplify that approach and test the same behavior. 

I added a test for:

 * normal snapshot taking and verify that all nodes have this snapshot
 * snapshotting triggers compaction verify log is compacted
 * snapshot is send on rejoin cluster

I plan to add further tests, more complex ones probably.

I checked #4467 and the problem was that the node was not able to truncate his log because it was at zero index and the leader was not able to send a snapshot, because it had none. The test was wrongly written I would say.

<!-- Please explain the changes you made here. -->

## Related issues

<!-- Which issues are closed by this PR or are related -->

closes #4467

## Pull Request Checklist

- [X] All commit messages match our [commit message guidelines](https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md#commit-message-guidelines)
- [X] The submitting code follows our [code style](https://github.com/zeebe-io/zeebe/wiki/Code-Style)
- [X] If submitting code, please run `mvn clean install -DskipTests` locally before committing

Zelldon commented:
@deepthidevaki @npepinpe someone has time to review this?

npepinpe commented on atomix/cluster/src/test/java/io/atomix/raft/RaftFailOverTest.java:
```diff
@@ -118,19 +117,97 @@ public void shouldRecoverLeaderRestart() throws Throwable {
   @Test
-  @Ignore("https://github.com/zeebe-io/zeebe/issues/4467")
-  public void testNodeCatchUpAfterCompaction() throws Exception {
+  public void shouldDoSnapshot() throws Exception {
```
Nit: snapshot is a verb, so just say `shouldSnapshot` ;)

Zelldon replied:
hm, but normally we say 'take a snapshot' right?

Zelldon commented:
> Generally looks good from my side, but there's a bunch of new things I don't know, do you think we could take 5 minutes to run through it together? sweat_smile

of course! when?

deepthidevaki commented on atomix/cluster/src/test/java/io/atomix/raft/RaftFailOverTest.java:
```diff
@@ -118,19 +117,97 @@ public void shouldRecoverLeaderRestart() throws Throwable {
+  }
+
+  @Test
+  public void shouldCompactLogOnSnapshot() throws Exception {
```
Isn't this testing the test setup rather than real implementation. Compaction is triggered by `RaftRule.RaftSnapshotListener`.

Zelldon replied:
I think you're totally right. But actually I would like to keep it, to make sure our test suite works also as expected. I could move it in a separate class called something like `RaftRuleTest` ? What do you think @deepthidevaki ?

deepthidevaki replied:
Do you really have to have the compaction logic in RaftRule? Isn't it used only in this test?

Zelldon replied:
Yes because with that I can test our bug we had, easily. Also other more complex scenarios

Zelldon replied:
Without this logic I wouldn't be able to test like shouldReplicateSnapshotOnJoin or shouldTakeMultipleSnapshotsAndReplicateSnapshotAfterRestart and also not what you mentioned below.

deepthidevaki replied:
Ok

deepthidevaki commented on atomix/cluster/src/test/java/io/atomix/raft/RaftFailOverTest.java:
```diff
@@ -118,19 +117,97 @@ public void shouldRecoverLeaderRestart() throws Throwable {
+    raftRule.doSnapshot(300);
+
+    // when
+    raftRule.startNode(follower);
```
In addition to this, another test that would be interesting is to restart follower with data loss. Can you add it?

Zelldon replied:
Of course!

Zelldon replied:
btw this should be similar to shouldReplicateSnapshotOnJoin

deepthidevaki replied:
It is different. When follower restarts with data loss, leader thinks that it has already send those events to the follower. There was a bug before - something like leader doesn't send the snapshot again if the follower restarts with data loss.

deepthidevaki commented on atomix/cluster/src/test/java/io/atomix/raft/RaftRule.java:
```diff
@@ -179,90 +187,119 @@ private MemberId nextNodeId() {
-      final RaftServer raftServer, final long index) {
-    raftServer.getContext().getServiceManager().setCompactableIndex(index);
-    return raftServer.compact();
+  public Snapshot snapshotOnNode(final String nodeId) {
```
```suggestion
  public Snapshot getSnapshotOnNode(final String nodeId) {
```

Zelldon commented:
Hey guys I incorporated your review hints and added additional tests for data loss and also to verify that entries after snapshots are only replicated.

@deepthidevaki  would be cool if you could take an additional look.

Zelldon commented:
Hey @deepthidevaki thanks I wanted to do that with this https://github.com/zeebe-io/zeebe/issues/4468

deepthidevaki commented:
Ok. I'm losing track of all refactoring issues :smile:

Zelldon commented:
No problem :smile: thanks for the review :+1:

Zelldon commented:
bors r+
